### ISSUE TITLE
Missed bounds simplifications

### ISSUE THREAD

**Issue by benoitsteiner** [halide/Halide#4521]
I tried the following code:
```cpp
  Halide::ImageParam input_a(Halide::type_of<float>(), 3, "input_a");
  input_a.dim(0).set_bounds(0, 1);
  Halide::ImageParam input_b(Halide::type_of<float>(), 3, "input_b");

  Halide::Func f("out_a"), g("out_b");
  Halide::Var i, j, k;
  f(i, j, k) = input_a(i, j, k) * 2;
  g(i, j, k) = input_b(i, j, k) / 2;

  std::vector<Halide::Func> fs;
  fs.push_back(f);
  fs.push_back(g);

  Halide::Pipeline p(fs);
  p.add_requirement(input_a.dim(0).extent() == input_b.dim(0).extent());
  p.add_requirement(input_a.dim(1).extent() == input_b.dim(1).extent());
  p.add_requirement(input_a.dim(2).extent() == 5);
  p.add_requirement(input_b.dim(2).extent() == 6);
  f.output_buffer().dim(1).set_bounds(0, 3);

  Halide::Target target("x86-64-linux-sse41-avx-avx2");

  std::vector<Halide::Argument> inputs;
  inputs.push_back(input_a);
  inputs.push_back(input_b);

 p.compile_to_lowered_stmt(std::string("/tmp/example.stmt"), inputs, Halide::Text, target);
```

I expected that all the extents of f and g be fully known. For f, the first extent would come from the bounds specification on input_a, the second from the explicit specification of the bounds of the output buffer, and the third from the pipeline requirement. For g, I was hoping that the pipeline requirements would also be leveraged to derive the bounds. However, in practice, the only known extent in example.stmt is the one specified through `f.output_buffer().dim(1).set_bounds(0, 3)`. All the other extents remain symbolic.

Of course in this trivial example I could directly set the bounds on f and g, but there are cases where this is difficult. Is there any way to improve the simplifier to start taking pipeline requirements and input extents into account ?

**Comment by abadams** [comment]
The simplifier does indeed exploit equality constraints, but f and g don't have a fixed size in your example - they can be anything up to the size of the input. So the asserts look like (output_size <= input_size), not (output_size == input_size).

**Comment by benoitsteiner** [comment]
Indeed, once I add requirements to enforce output_size == input_size the loop extents are inferred as I expected. The only thing that still missing is to decouple the simplifiers from the lowering process in order to ensure that we can call them before starting the autoscheduler.

**Comment by abadams** [comment]
> [!NOTE]
> **Automated triage:** this issue was flagged by Claude as likely safe to close, for the
> reason below. The reasoning was checked against the current source, and a second pass
> tried and failed to refute it, but no human verified this specific text.
> **If the reason is wrong, please reopen the issue and say so** — some fraction of these
> are expected to be wrong.

**Reason (CLOSE_WONTFIX):** The thread resolved the reported behavior: the bounds are not inferred because the pipeline's asserts are inequalities (output_size \<= input_size), not equalities, and the reporter confirmed that adding equality requirements makes the extents constant-fold as expected. The only residue is an unscoped wish to 'decouple the simplifiers from lowering so they can run before the autoscheduler'.

**Evidence:** Issue thread: abadams 2020-01-10 explains the \<= vs == asserts; benoitsteiner 2020-01-10 confirms 'once I add requirements to enforce output_size == input_size the loop extents are inferred as I expected'.

Comment and close: 'Resolved in-thread — the extents stay symbolic because the injected asserts are inequalities; adding `==` requirements makes them fold, as you confirmed. The remaining idea (running bounds simplification before the autoscheduler) is a separate, much larger design question — please open a focused issue if you still want it.'
